// azure-event-hubs/dist/lib/connectionContext.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const debugModule = require("debug");
const uuid = require("uuid/v4");
const Constants = require("./util/constants");
const _1 = require(".");
const managementClient_1 = require("./managementClient");
const cbs_1 = require("./cbs");
const sas_1 = require("./auth/sas");
const dataTransformer_1 = require("./dataTransformer");
const debug = debugModule("azure:event-hubs:connectionContext");
var ConnectionContext;
(function (ConnectionContext) {
    /**
     * @property {string} userAgent The user agent string for the event hub client. Constant value: "/js-event-hubs".
     */
    ConnectionContext.userAgent = "/js-event-hubs";
    function create(config, options) {
        _1.ConnectionConfig.validate(config);
        if (!options)
            options = {};
        const context = {
            connectionLock: `${Constants.establishConnection}-${uuid()}`,
            negotiateClaimLock: `${Constants.negotiateClaim}-${uuid()}`,
            config: config,
            tokenProvider: options.tokenProvider ||
                new sas_1.SasTokenProvider(config.endpoint, config.sharedAccessKeyName, config.sharedAccessKey),
            senders: {},
            receivers: {},
            dataTransformer: options.dataTransformer || new dataTransformer_1.DefaultDataTransformer()
        };
        context.cbsSession = new cbs_1.CbsClient(context);
        const mOptions = {
            address: options.managementSessionAddress,
            audience: options.managementSessionAudience
        };
        context.managementSession = new managementClient_1.ManagementClient(context, mOptions);
        debug("Created connection context: %O", context);
        return context;
    }
    ConnectionContext.create = create;
})(ConnectionContext = exports.ConnectionContext || (exports.ConnectionContext = {}));
